### PR TITLE
V2 dev

### LINKED ISSUE #69: Use tracksdata instead of networkx as the backend
### Things I know we will need to change in funtracks:
 - Add tracksdata dependency
 - Tracks object
     - instead of having a networkx graph, they will have a tracksdata graph
     - remove the segmentation
 - Any actions that update the segmentation need to use the tracksdata graph API to update them
 - Anywhere else we use networkx outside of Tracks (which shouldn't happen but does), we replace with tracksdata API calls
 
### Things we will need to change in finn:
- TrackLabels uses tracksdata GraphArrayView instead of np.array for the segmentation
... infinite colormap hell, presumably 🤷‍♀️ 

### Need:
- Test graph in sql

### PR DESCRIPTION
**Pull request by cmalinmayor**
- [x] 1. funtracks: tracksdata merge into v2-dev (#126) - Teun!
- [x] 2. funtracks: Teun!
          - https://github.com/funkelab/funtracks/issues/149
          - https://github.com/funkelab/funtracks/issues/156
          - https://github.com/funkelab/funtracks/issues/155 
- [x] 3. MT: update Motile Tracker to work on main with no deprecation warnings 
          - #132 
          - https://github.com/funkelab/motile_tracker/pull/330
- [x] 4. funtracks: Merge main into v2-dev (this PR)
- [x] 5. test motile tracker on v2-dev and fix issues - Teun and Anniek
          - https://github.com/funkelab/motile_tracker/pull/355 

Closes #69 #142

Visual overview:
Screenshot 2026-01-29 at 16 02 44

### CODE REVIEW

**Comment by TeunHuijben** [comment]
This complicated todo-list needed a diagram very badly (at least my head did 😂). Please correct me if any step/arrow is wrong or incomplete

File: src/funtracks/actions/add_delete_edge.py [hunk 1]
```diff
@@ -52,7 +54,24 @@ def _apply(self) -> None:
                     f"Cannot add edge {self.edge}: endpoint {node} not in graph yet"
                 )
 
-        self.tracks.graph.add_edge(self.edge[0], self.edge[1], **self.attributes)
+        if self.tracks.graph.has_edge(*self.edge):
+            raise ValueError(f"Edge {self.edge} already exists in the graph")
+
+        # Add required solution attribute
```
Comment (cmalinmayor) [inline src/funtracks/actions/add_delete_edge.py]: Does this work with undo? Or do we need to add the solution attr to self.attributes?

File: src/funtracks/candidate_graph/utils.py [hunk 2]
```diff
@@ -1,88 +1,116 @@
...
 
 logger = logging.getLogger(__name__)
 
 
 def nodes_from_segmentation(
     segmentation: np.ndarray,
     scale: list[float] | None = None,
-) -> tuple[nx.DiGraph, dict[int, list[Any]]]:
-    """Extract candidate nodes from a segmentation. Returns a networkx graph
+    mask: bool = False,
```
Comment (cmalinmayor) [inline src/funtracks/candidate_graph/utils.py]: I think we can default this to True?

File: src/funtracks/candidate_graph/utils.py [hunk 3]
```diff
@@ -1,88 +1,116 @@
...
+                "pos": list(regionprop.centroid),
+                "area": float(regionprop.area),
+            }
+            if mask:
+                # regionprop.image is the cropped boolean mask within the bbox,
+                # already computed by regionprops at no extra cost.
+                # regionprop.bbox format: (min_0, ..., min_n, max_0, ..., max_n),
+                # which matches the Mask bbox convention.
+                bbox = np.array(regionprop.bbox, dtype=np.int64)
+                attrs["mask"] = Mask(np.ascontiguousarray(regionprop.image), bbox=bbox)
```
Comment (cmalinmayor) [inline src/funtracks/candidate_graph/utils.py:92]: Why is the bbox stored twice, once on mask and once on it's own attr?

File: src/funtracks/data_model/tracks.py [hunk 4]
```diff
@@ -57,23 +60,21 @@ class Tracks:
...
+        _segmentation: GraphArrayView | None = None,
     ):
         """Initialize a Tracks object.
 
         Args:
-            graph (nx.DiGraph): NetworkX directed graph with nodes as detections and
-                edges as links.
-            segmentation (np.ndarray | None): Optional segmentation array where labels
-                match node IDs. Required for computing region properties (area, etc.).
+            graph (td.graph.GraphView): NetworkX directed graph with nodes as detections
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py]:
```suggestion
            graph (td.graph.GraphView): tracksdata directed graph with nodes as detections
```

File: src/funtracks/data_model/tracks.py [hunk 5]
```diff
@@ -321,7 +332,8 @@ def _setup_core_computed_features(self) -> None:
         for annotator in self.annotators:
             if isinstance(annotator, RegionpropsAnnotator):
                 pos_key = annotator.pos_key
-                self.features.position_key = pos_key
+                if self.features.position_key is None:
+                    self.features.position_key = pos_key
                 core_computed_features.append(pos_key)
                 # special case for backward compatibility
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py:338]: This is our opportunity to stop auto-computing area as a special backward compatible feature - just worth considering before we release, not necessary to merge this PR

File: src/funtracks/data_model/tracks.py [hunk 6]
```diff
@@ -461,136 +515,27 @@ def get_time(self, node: Node) -> int:
...
-            "`get_iou` is deprecated and will be removed in funtracks v2.0. "
-            "Use `get_edge_attr(edge, 'iou')` instead.",
-            DeprecationWarning,
-            stacklevel=2,
-        )
-        return self.get_edge_attr(edge, "iou")
+        mask = self.graph.nodes[node][td.DEFAULT_ATTR_KEYS.MASK]
+        return mask
 
     def get_pixels(self, node: Node) -> tuple[np.ndarray, ...] | None:
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py:540]: Do we need this anymore?

File: src/funtracks/data_model/tracks.py [hunk 7]
```diff
@@ -606,55 +551,61 @@ def get_pixels(self, node: Node) -> tuple[np.ndarray, ...] | None:
...
-        return (time_array, *loc_pixels)
 
-    def set_pixels(self, pixels: tuple[np.ndarray, ...], value: int) -> None:
-        """Set the given pixels in the segmentation to the given value.
+        # Convert to pixels
+        pixels = td_mask_to_pixels(mask, time, ndim=self.ndim)
+
+        return pixels
+
+    def _update_segmentation_cache(self, mask: td.Mask, time: int) -> None:
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py:564]: Should this logic go into td at some point?

File: src/funtracks/data_model/tracks.py [hunk 8]
```diff
@@ -723,54 +674,37 @@ def _compute_ndim(
     def _set_node_attr(self, node: Node, attr: str, value: Any):
         if isinstance(value, np.ndarray):
             value = list(value)
-        self.graph.nodes[node][attr] = value
+        self.graph.nodes[node][attr] = [value]
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py:677]: The Tracksdata API requiring this to be a list is a little weird, but I can accept it.

File: src/funtracks/data_model/tracks.py [hunk 9]
```diff
@@ -837,112 +771,55 @@ def disable_features(self, feature_keys: list[str]) -> None:
...
-            DeprecationWarning,
-            stacklevel=2,
-        )
-        from ..import_export.internal_format import delete_tracks
+        # Add to the features dictionary
+        self.features[key] = feature
+
+        # Perform custom graph operations when a feature is added
+        if feature["feature_type"] == "node" and key not in self.graph.node_attr_keys():
+            dtype = to_polars_dtype(feature["value_type"])
```
Comment (cmalinmayor) [inline src/funtracks/data_model/tracks.py:791]: Where does polars come into our dependencies? And why? Just curious

File: src/funtracks/annotators/_compute_ious.py [hunk 10]
```diff
@@ -1,35 +1,15 @@
-import numpy as np
+from tracksdata.nodes._mask import Mask
 
 
-def _compute_ious(frame1: np.ndarray, frame2: np.ndarray) -> list[tuple[int, int, float]]:
-    """Compute label IOUs between two label arrays of the same shape. Ignores background
-    (label 0).
+def _compute_iou(mask1: Mask, mask2: Mask) -> list[tuple[int, int, float]]:
```
Comment (cmalinmayor) [inline src/funtracks/annotators/_compute_ious.py]: We can delete this function lol

File: src/funtracks/annotators/_regionprops_annotator.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 from typing import TYPE_CHECKING, NamedTuple
 
 import numpy as np
+from tracksdata.nodes._mask import Mask
```
Comment (cmalinmayor) [inline src/funtracks/annotators/_regionprops_annotator.py]: Can tracksdata expose types in the public API?

File: src/funtracks/annotators/_track_annotator.py [hunk 12]
```diff
@@ -207,19 +218,40 @@ def _assign_tracklet_ids(self) -> None:
         After removing division edges, each connected component will get a unique ID,
         and the relevant class attributes will be updated.
         """
-        graph_copy = self.tracks.graph.copy()
-        parents = [node for node, degree in self.tracks.graph.out_degree() if degree >= 2]
+        graph_copy = self.tracks.graph.detach().filter().subgraph()
+
+        parents = [
```
Comment (cmalinmayor) [inline src/funtracks/annotators/_track_annotator.py:223]: Yet another way to find parent nodes 🙃 we really need to make a function for this

File: src/funtracks/annotators/_track_annotator.py [hunk 13]
```diff
@@ -207,19 +218,40 @@ def _assign_tracklet_ids(self) -> None:
...
+                self.tracks.graph.node_ids(), self.tracks.graph.out_degree(), strict=True
+            )
+            if degree >= 2
+        ]
 
         # Remove all intertrack edges from a copy of the original graph
         for parent in parents:
-            daughters = self.tracks.successors(parent)
+            all_edges = self.tracks.graph.edge_list()
+            daughters = [edge[1] for edge in all_edges if edge[0] == parent]
```
Comment (cmalinmayor) [inline src/funtracks/annotators/_track_annotator.py:234]: Is successors really so much worse than looping in python? Even the sql has to be be faster than this, no?

File: src/funtracks/actions/add_delete_edge.py (hunk 1 above)
Comment (TeunHuijben) [reply src/funtracks/actions/add_delete_edge.py]: Good point! When an edge is added, it will always have `solution==True`, because the user wants the edge. Deleting an edge simply deletes the edge and saved the attributes. 
Since I don't see a scenario where a user will delete an edge from the candidate graph (`solution==False`), every deleted edge that will ever be undone, will have `solution==True`. But I agree, it is cleaner to add it to the `self.attributes` to be completely sure

File: src/funtracks/annotators/_compute_ious.py (hunk 10 above)
Comment (TeunHuijben) [reply src/funtracks/annotators/_compute_ious.py]: done

File: src/funtracks/annotators/_regionprops_annotator.py (hunk 11 above)
Comment (TeunHuijben) [reply src/funtracks/annotators/_regionprops_annotator.py]: Good call, done

File: src/funtracks/annotators/_track_annotator.py (hunk 12 above)
Comment (TeunHuijben) [reply src/funtracks/annotators/_track_annotator.py:223]: yes, this is extremely inefficient 😂!!! There is a current funtracks issue to refactor the track_id/lineage_id logic, and use the efficient tracksdata equivalent functions. I'll close this comment in regard of that issue

File: src/funtracks/annotators/_track_annotator.py (hunk 13 above)
Comment (TeunHuijben) [reply src/funtracks/annotators/_track_annotator.py:234]: Same here, refering to isssue [#148](https://github.com/funkelab/funtracks/issues/148)

File: src/funtracks/candidate_graph/utils.py (hunk 3 above)
Comment (TeunHuijben) [reply src/funtracks/candidate_graph/utils.py:92]: That is a tracksdata decision for efficiency. I adds a very minor extra memory, but in exchange the loading is a lot more efficient: 
- saving the bboxes separately makes it fast to look up overlaps/neighbors etc, without loading the entire mask
- when in need of the actual mask, loading the `Mask` object is enough, because it also contains the bbox. It prevents you from having to load the bbox attribute as well

File: src/funtracks/candidate_graph/utils.py (hunk 2 above)
Comment (TeunHuijben) [reply src/funtracks/candidate_graph/utils.py]: done

File: src/funtracks/data_model/tracks.py (hunk 5 above)
Comment (TeunHuijben) [reply src/funtracks/data_model/tracks.py:338]: made an issue

File: src/funtracks/data_model/tracks.py (hunk 6 above)
Comment (TeunHuijben) [reply src/funtracks/data_model/tracks.py:540]: Fair point, it is only used in the tests, but never in the real code. We can keep it as a "convenience function" for the tests, otherwise all tests (20 occurences) need to be changed. It is probably best to remove it and reformat the tests, I made an issue: #194

File: src/funtracks/data_model/tracks.py (hunk 7 above)
Comment (TeunHuijben) [reply src/funtracks/data_model/tracks.py:564]: Good point! Actually, it is already part of td! I'll make an issue to remove it here, good catch 🙌

File: src/funtracks/data_model/tracks.py (hunk 8 above)
Comment (TeunHuijben) [reply src/funtracks/data_model/tracks.py:677]: I'll look into this! Made an issue

File: src/funtracks/data_model/tracks.py (hunk 9 above)
Comment (TeunHuijben) [reply src/funtracks/data_model/tracks.py:791]: Yeah, bit messy, I agree. But we need it every time we add a new node|edge attribute key, because it requires defining the type. I will ask Jordao, whether this cannot be moved into td. Made an issue
